### PR TITLE
emit_x64: Refactor patching code

### PR DESCRIPTION
**Pull request by MerryMage**
Changes:

* Only have a single std::unordered_map for patching information
* Factor patch emitters into own functions
* Implement EmitX64::Unpatch

Ping @lynn.

### CODE REVIEW

File: src/backend_x64/emit_x64.cpp [hunk 1]
```diff
@@ -2775,38 +2772,63 @@ void EmitX64::EmitTerminalCheckHalt(IR::Term::CheckHalt terminal, IR::LocationDe
...
+    if (target_code_ptr) {
+        code->jmp(target_code_ptr);
+    } else {
+        code->mov(MJitStateReg(Arm::Reg::PC), target_desc.PC());
+        code->jmp(code->GetReturnFromRunCodeAddress());
+    }
+    code->EnsurePatchLocationSize(patch_location, 13);
+}
+
+void EmitX64::EmitPatchPushRsb(CodePtr target_code_ptr) {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: A better name would be `EmitPatchMovRcx`

File: src/backend_x64/emit_x64.h [hunk 2]
```diff
@@ -71,7 +71,18 @@ class EmitX64 final {
     void EmitTerminalPopRSBHint(IR::Term::PopRSBHint terminal, IR::LocationDescriptor initial_location);
     void EmitTerminalIf(IR::Term::If terminal, IR::LocationDescriptor initial_location);
     void EmitTerminalCheckHalt(IR::Term::CheckHalt terminal, IR::LocationDescriptor initial_location);
-    void Patch(IR::LocationDescriptor desc, CodePtr bb);
+
+    // Patching
+    struct PatchInformation {
+        std::vector<CodePtr> jg;
+        std::vector<CodePtr> jmp;
+        std::vector<CodePtr> push_rsb;
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.h]: Prefer `mov_rcx`.

File: src/backend_x64/emit_x64.cpp [hunk 3]
```diff
@@ -2775,38 +2772,63 @@ void EmitX64::EmitTerminalCheckHalt(IR::Term::CheckHalt terminal, IR::LocationDe
     EmitTerminal(terminal.else_, initial_location);
 }
 
-void EmitX64::Patch(IR::LocationDescriptor desc, CodePtr bb) {
-    using namespace Xbyak::util;
-
+void EmitX64::Patch(const IR::LocationDescriptor& desc, CodePtr bb) {
     const CodePtr save_code_ptr = code->getCurr();
+    const PatchInformation patch_info = patch_information[desc.UniqueHash()];
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `const PatchInformation&`

**Comment by FernandoS27** [comment]
hmn so this would make blocks position independent or there's still a need to identify and patch modifications to the PC?

**Comment by MerryMage** [comment]
@FernandoS27 There would always been a need to patch whether or not the blocks are position independent.

**Pull request merged**